QuantConnect.GDAXBrokerage/Api/CoinbaseApi.cs

```diff
@@ -0,0 +1,326 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using RestSharp;
+using System.Linq;
+using Newtonsoft.Json;
+using QuantConnect.Util;
+using QuantConnect.Orders;
+using System.Globalization;
+using QuantConnect.Brokerages;
+using QuantConnect.Securities;
+using System.Collections.Generic;
+using QuantConnect.CoinbaseBrokerage.Models;
+using QuantConnect.CoinbaseBrokerage.Converters;
+using QuantConnect.CoinbaseBrokerage.Models.Enums;
+using QuantConnect.CoinbaseBrokerage.Models.Requests;
+using BrokerageEnums = QuantConnect.CoinbaseBrokerage.Models.Enums;
+
+namespace QuantConnect.CoinbaseBrokerage.Api;
+
+public class CoinbaseApi : IDisposable
+{
+    /// <summary>
+    /// Represents the maximum number of occurrences allowed per unit of time for a gate limit.
+    /// </summary>
+    /// <remarks>
+    /// Refer to the documentation for more details: 
+    /// <see href="https://docs.cloud.coinbase.com/advanced-trade-api/docs/rest-api-rate-limits"/>.
+    /// </remarks>
+    private const int maxGateLimitOccurrences = 30;
+
+    /// <summary>
+    /// Represents an instance of the Coinbase API client used for communication with the Coinbase API.
+    /// </summary>
+    private readonly CoinbaseApiClient _apiClient;
+
+    /// <summary>
+    /// Represents the prefix used for API endpoints in the application.
+    /// </summary>
+    private readonly string _apiPrefix = "/api/v3";
+
+    private JsonSerializerSettings _jsonSerializerSettings = new()
+    {
+        Converters = new List<JsonConverter>() { new CoinbaseDecimalStringConverter() },
+        NullValueHandling = NullValueHandling.Ignore
+    };
+
+    /// <summary>
+    /// Symbol mapper
+    /// </summary>
+    private ISymbolMapper SymbolMapper { get; }
+
+    /// <summary>
+    /// Security provider
+    /// </summary>
+    private ISecurityProvider SecurityProvider { get; }
+
+    public CoinbaseApi(ISymbolMapper symbolMapper, ISecurityProvider securityProvider,
+        string apiKey, string apiKeySecret, string restApiUrl)
+    {
+        SymbolMapper = symbolMapper;
+        SecurityProvider = securityProvider;
+        _apiClient = new CoinbaseApiClient(apiKey, apiKeySecret, restApiUrl, maxGateLimitOccurrences);
+    }
+
+    public (string apiKey, string timestamp, string signature) GetWebSocketSignatures(string channel, ICollection<string> productIds)
+    {
+        return _apiClient.GenerateWebSocketSignature(channel, productIds);
+    }
+
+    /// <summary>
+    /// Retrieves a list of Coinbase accounts associated with the authenticated user's brokerage.
+    /// </summary>
+    /// <returns>An IEnumerable of CoinbaseAccount objects representing the user's brokerage accounts.</returns>
+    public IEnumerable<CoinbaseAccount> GetListAccounts()
+    {
```

Comment:
Can this and the rest of the requests return in a failure response? If so, would the JSON deserialization fail?

Reply:
Added handler for bad http response